### PR TITLE
Fix Store::retrieve()

### PR DESCRIPTION
**Pull request by neithernut**
This PR targets a primary issue laid out in #103.

### CODE REVIEW

**Comment by neithernut** [comment]
Note: the way the store entry status is currently updated is inadequate. Maybe I'll come up with a better solution, maybe I'll move the contents of `get_entry()` to another place.

**Comment by matthiasbeyer** [comment]
@neithernut Have a look at my branch https://github.com/matthiasbeyer/imag/tree/pr-145-cleanup-suggestions , maybe you want to ff-merge my suggestions? It was hacketyhackhack and it compiles, though I didn't review myself.

**Comment by neithernut** [comment]
Apparently, it doesn't compile. I'll have a look at it later.

**Comment by matthiasbeyer** [comment]
ah. Remove the 0443a0d commit, this breaks the build.

**Comment by neithernut** [comment]
I thought so, but I was not sure how important forwarding the error was to you. Personally, I think that panicking is acceptable if the store itself (or a locking mechanism within the store) is broken.

**Comment by matthiasbeyer** [comment]
Forwarding it would be nice, though I don't mind it that much, as the callee can rather clearly see what happened here. Panicing isn't an option as we are writing a library. I guess you mean returning an error.

**Comment by neithernut** [comment]
No, I meant panicking. But in this case, the `delete()` method must also be adjusted.

**Comment by TheNeikos** [comment]
No, you don't want to forward that error.

**Comment by matthiasbeyer** [comment]
@TheNeikos ACK here?

File: libimagstore/src/store.rs [hunk 1]
```diff
@@ -142,15 +142,16 @@ impl Store {
...
-            .and_then(|store_entry| store_entry.get_entry())
-            .and_then(|entry| Ok(FileLockEntry::new(self, entry, id)))
-    }
+        self.entries
+            .write()
+            .map_err(|_| StoreError::new(StoreErrorKind::LockPoisoned, None))
+            .and_then(|mut es| {
+                let mut se = es.entry(id.clone()).or_insert_with(|| StoreEntry::new(id.clone()));
+                se.status = StoreEntryStatus::Borrowed;
+                se.get_entry()
```
Comment (TheNeikos) [inline libimagstore/src/store.rs]: This will always fail because it is already borrowed at this point.

**Comment by TheNeikos** [comment]
Nah I don't ACK this, and I didn't have the time to write up an answer to the previous critique which is valid in the current context but is easily explained because the code itself is in an intermediary state.

File: libimagstore/src/store.rs (hunk 1 above)
Comment (neithernut) [inline libimagstore/src/store.rs]: Yep, he's right. Now I also remember why I had that many bindings.

**Comment by neithernut** [comment]
So, you generally won't ACK this PR (and come up with a fix of your own)? Or can you point out concrete things you want to have changed?

**Comment by TheNeikos** [comment]
After discussing with @matthiasbeyer what he would like to see in the code, which is radically different than what I envisioned, I can ACK this.

**Comment by neithernut** [comment]
Done.

**Comment by matthiasbeyer** [comment]
AFAIK, @TheNeikos doesn't consider this ready.

**Comment by TheNeikos** [comment]
I'm fine with this.

**Comment by matthiasbeyer** [comment]
So you want me to merge this?

**Comment by TheNeikos** [comment]
Aye

**Pull request merged**